app/services/puppet_fact_parser.rb

```diff
@@ -68,7 +68,7 @@ def operatingsystem
   def environment
     # by default, puppet doesn't store an env name in the database
     name = facts[:environment] || facts[:agent_specified_environment] || Setting[:default_puppet_environment]
-    Environment.where(:name => name).first_or_create
+    Environment.unscoped.where(:name => name).first_or_create
```

Comment:
When using unscoped here but being in a taxonomy context, the "first" part will match without the taxonomy (so we don't get errors if the taxonomies don't match) but the "create" part will still take into account the taxonomy context and assign it to the new object.